web_src/js/features/comp/EditorUpload.ts

```diff
@@ -99,8 +101,8 @@ async function handleUploadFiles(editor, dropzoneEl, files, e) {
     const {width, dppx} = await imageInfo(file);
     const placeholder = `[${name}](uploading ...)`;
 
-    editor.insertPlaceholder(placeholder);
-    await uploadFile(dropzoneEl, file); // the "file" will get its "uuid" during the upload
+    // the "file" will get its "uuid" during the upload
+    await uploadFile(dropzoneEl, file, () => editor.insertPlaceholder(placeholder));
```

Comment:
```suggestion
    // defer the insertion to next tick, then onContentChanged could read "isUploading=true" correctly
    setTimeout(0, () => editor.insertPlaceholder(placeholder));
    // the "file" will get its "uuid" during the upload
    await uploadFile(dropzoneEl, file, () => editor.insertPlaceholder(placeholder));
```

Reply:
I just think again about the problem. It looks that the `setTimeout` is somewhat hacky, and it doesn't guarantee that it would work wither other uploaders.

I will try to make some new changes to introduce a separate event to handle this situation.

Reply:
In this suggestion, third parameter (placeholder) seems not to be necessary. isn't it?

Reply:
Yup, the first suggestion is to remove the `placeholderCallback`.

But indeed the `setTimeout` seems somewhat hacky, and if we'd like to improve the uploader or change it, it might become a blocker.

----

So, a more general solution in my mind is: introduce a more general "opts.onUploadStateChanged" callback, make the ComboMarkdownEditor listen on Dropzone events and call the onUploadStateChanged, then UI code could sync the UI state accordingly. It could 100% isolate the business code and the uploader and be clearer.

Reply:
Well, one more thing ........ the current solution (onContentChanged) is also buggy when the user uploads a file by the Dropzone directly, there is no content change, then there is no "isUploading" check .....

Quote my previous comment:

> I think the UI states should be managed by some well-defined events&handlers, but not by an ad-hoc "thumbnail" event with onContentChanged callback.


So, the problem is not that simple as it looks. I will take a look when I have time.